### DISCUSSION TITLE
How to deal with a remote translation?

### DISCUSSION THREAD

**Discussion by NurdinDev** [vinissimus/next-translate#451] [Q&A]
I'm dealing with a CMS that gives me all the translation dynamically via API, how to deal with this type of translation file if the file does not exist in the local directory?  

My mind telling me to fetch the translation JSON as is, and find a way to save it in my local directory, but I don't have the right answer for this case. 

Thanks in advance.

**Comment by aralroca** [comment]
Did you try something like this?

```js
const fetch = require('isomorphic-unfetch')
// ...
{
  // ...rest of config
  "loadLocaleFrom": (lang, ns) =>
    fetch(`${API_URL}/myTranslationsFiles/${lang}/${ns}`).then((r) => r.json()),
}
```

**Reply by NurdinDev** [reply]
```
const fetch = require('isomorphic-unfetch');

module.exports = {
  "locales": ["en-US", "tr"],
  "defaultLocale": "en-US",
  "loadLocaleFrom": (lang, ns) =>
    fetch('https://myapi...../api/words').then((r) => {
      console.log("-----------------------");
      console.log(r.json());
      const json = r.json();
      return json[lang];
    })
}
```
and I used  `{ t } = useTranslation()` on the page.

But nothing happens even no console log printed from `loadLocaleFrom`.

**Reply by aralroca** [reply]
You miss the `pages` on the [configuration](https://github.com/vinissimus/next-translate/tree/1.0.1#3-configuration). This is mandatory for each page to know which namespaces to use. If every page uses the same one, you can do:

```diff
const fetch = require('isomorphic-unfetch');

module.exports = {
  "locales": ["en-US", "tr"],
  "defaultLocale": "en-US",
  "loadLocaleFrom": (lang, ns) =>
    fetch('https://myapi...../api/words').then((r) => {
      console.log("-----------------------");
      console.log(r.json());
      const json = r.json();
      return json[lang];
+    }),
+  "pages": {"*": ["common"]}
}
```

Example: https://github.com/vinissimus/next-translate/blob/1.0.1/examples/basic/i18n.json

**Reply by NurdinDev** [reply]
And what about if I don't have namespaces If I'm just fetching JSON words?

**Reply by aralroca** [reply]
Each page looks at what namespaces it has and makes a request for each namespace. So it is mandatory to put a namespace. even if it is a unique identifier for all pages...

If you have a page without namespaces then no request will be made for this page.

**Reply by NurdinDev** [reply]
Sorry, I got lost in the namespaces concept
is it a file structure or it can be a JSON structure?
JSON like: 
```
common: {
 key: value
}
```
or file like:
`common.json`

**Reply by aralroca** [reply]
Namespaces are inspired by this feature of i18next: https://www.i18next.com/principles/namespaces

As default, they work like `/locales/en/[namespace].json`. Ex: `/locales/en/product.json`, `/locales/en/checkout.json`, etc. This way on each page you can indicate which namespaces it needs, so it hasn't to download all the translations.

You can make it work as `common: { key: value }` overwriting the `loadLocaleFrom` to something like:

```js
"loadLocaleFrom": async (lang, ns) =>
    const translations = await import(`./locales/${lang}.json`).then(m => m.default)
    return translations[ns]
}),
```

However, I recommend using the default one so that you only dynamically download the namespace you need.

**Reply by NurdinDev** [reply]
I created a node script to fetch my JSON words and store them inside the locales directory :D

**Reply by nraptis1978** [reply]
I have 2 questions:
1. What if I want to update the translations from my CMS. How the json files will be updated ? I see that you created a script the updates the files. how do you run this script ?
2. When using the translations on a page that have SSG I don't think that the content will be updated with new translations even if the locales/en .... json files are changed

**Reply by aralroca** [reply]
It's using the loader of the page. SSG -> is calling `loadLocaleFrom` inside `getStaticProps`. SSR -> is calling `loadLocaleFrom` inside `getInitialProps` / `getServerSideProps`.

https://github.com/vinissimus/next-translate/blob/a68ad2b9b87be627b7024f8e35867c444ff439d8/src/plugin/templateWithLoader.ts#L91

Is updated if you provide the Next.js `revalidate` prop: https://nextjs.org/docs/api-reference/data-fetching/get-static-props#revalidate

**Reply by nraptis1978** [reply]
Hi, I've tested this and seems to be working fine with getStaticProps and revalidate on a page. I'm using
 loadLocaleFrom: (lang, ns) =>  fetch(`${apiUrl}/${lang}`).then((res) => res.json());
    
What if I have 20 pages with revalidate = "1 hour" and they share the same translations and there is no namespace. I mean that I have 1 api call for all translations of all 20 pages per lang. If I add getStaticProps and revalidate on all pages it will trigger 20 times the same api call per hour ( or not ? ). Can somehow make this to 1 api call ? i.e. update the translations only in 1 page and share it across all others ?

**Reply by nraptis1978** [reply]
> I created a node script to fetch my JSON words and store them inside the locales directory :D

I've try this and even though I have in my pages getStaticProps with revalidate and on i18n.js this:
  loadLocaleFrom: (lang, ns) =>
    import(`./locales/${lang}/common.json`).then((m) => m.default),
    
when I'm updating the translations inside the common.json there is no update in the pages...
